torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -269,14 +250,33 @@ void ProcessGroupAgent::sync() {
 }
 
 void ProcessGroupAgent::start() {
+  rpcRunning_.store(true);
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
   futureTimeoutThread_ =
       std::thread(&ProcessGroupAgent::pollTimedOutRPCs, this);
 }
 
+void ProcessGroupAgent::shutdown() {
+  if (!rpcRunning_.exchange(false)) {
+    return;
+  }
+  LOG(INFO) << "Stopping ProcessGroupAgent.";
+  {
+    std::unique_lock<std::mutex> lock(recvWorkMutex_);
+    if (recvWork_) {
+      recvWork_->abort();
+    }
+  }
+  threadPool_.waitWorkComplete();
```

Comment:
I had to add this here in shutdown, otherwise the test `test_rpc_shutdown()` would flake with gloo connection reset errors. I don't think it delays the shutdown by too much, as it just waits for already enqueued work to complete.

Reply:
hmm, does this need to be placed before abort? Otherwise, if there are indeed unfinished task, the receiving end listener thread might have already been aborted?

Reply:
Hm, my reasoning for doing it after was that if it was before, we could possibly get more work while aborting the listenerThread, and then enqueue that recv, and not wait for it to be completed. For example, say that recvWork->wait has already been unblocked before we call its abort (via a message from another worker), we may have called enqueueRecv which will add a task to the thread pool.